Add mask command and modprobe test with blacklist

howardthuang opened the pull request:
Closes: https://github.com/kmod-project/kmod/issues/40

libkmod: add mask command

Co-authored-by: Jakub Ślepecki <jakub.slepecki@intel.com>
Signed-off-by: Dawid Osuchowski <dawid.osuchowski@intel.com>
Signed-off-by: Jakub Ślepecki <jakub.slepecki@intel.com>

testsuite: test-modprobe: modprobe with blacklist

Co-authored-by: Dawid Osuchowski <dawid.osuchowski@intel.com>
Signed-off-by: Jakub Ślepecki <jakub.slepecki@intel.com>
Signed-off-by: Dawid Osuchowski <dawid.osuchowski@intel.com>

whythefyouflying commented:
PR Title:
Introduce mask command to prevent modules from loading
 
PR Description:
The behavior of the new mask command could easily be replaced by modprobe -b in downstream, hence the first commit. Otherwise this allows user to prevent loading selected modules without changing any already established behaviour.

jspcki commented:
To add more context to our initial description. In PR there are three options to prevent module from loading:

- install (used with `modprobe alias_or_module`)
- blacklist (used with `modprobe alias` and `modprobe -b module`)
- mask (used with `modprobe alias_or_module`)

Install still provides the best flexibility: user may specify either `/bin/false` or `/bin/true` (and a lot more complex options) to get different resulting behaviour out of modprobe.

Blacklist puts the most burden on the user. If they choose to disable module with it (in contrast to just limiting the alias resolution), they need to make sure their scripts use either aliases or -b.

Mask intends to provide a strict (in contrast to install) and clear intent-based (in contrast to both blacklist and install) option that
does not change any currently established behaviour. This way we also avoid the problem of deprecation of install.

whythefyouflying commented:
The latest force-push was to address the formatting issue reported by CI: https://github.com/kmod-project/kmod/actions/runs/13732774455/job/38592878799?pr=312

evelikov reviewed (changes requested):
Overall I'm loving how you went above and beyond on the prep (blacklist tests) patch. There's one serious note, but the rest are cosmetic suggestions.

Thanks team o/

whythefyouflying commented on libkmod/libkmod-config.c:
```diff
@@ -803,6 +834,13 @@ static int kmod_config_parse(struct kmod_config *config, int fd, const char *fil
+			if (underscores(modname) < 0)
+				goto syntax_error;
+
+			kmod_config_add_mask(config, modname);
```
Will do across the patch

Jakub is willing to add to that and sort across the tree

whythefyouflying replied:
> This one is done. I don't think we adjusted it across the tree, @whythefyouflying?

It was not. I believe that would best be suited in a separate series/PR.

whythefyouflying commented on libkmod/libkmod-config.c:
```diff
@@ -615,13 +641,18 @@ static void kcmdline_parse_result(struct kmod_config *config, char *modname, cha
 	DBG(config->ctx, "%s %s\n", modname, param);
 
-	if (streq(modname, "modprobe") && !strncmp(param, "blacklist=", 10)) {
+	if (streq(modname, "modprobe")) {
```
Agreed, will change

whythefyouflying replied:
Done.

jspcki commented on testsuite/test-modprobe.c:
```diff
@@ -406,4 +416,19 @@ DEFINE_TEST(modprobe_blacklisted_by_name_filtered,
+	},
+	.modules_loaded = "",
+	.modules_not_loaded = "mod-simple",
+	);
```
Agreed, I'd say:

* dependency - module does not load, modprobe status is non-zero
* softdep - target module loads without softdep(s), status is zero
* weakdep - target module loads without weakdep(s), status is zero
 
I added such tests in local branch and they will be included in v3. I settled on weakdep behaviour because of its description in manual pages "Those are similar to pre softdep" and "userspace doesn't attempt to load that dependency before the specified module."

This is not what's happening right now, so we'll fix it.

jspcki replied:
Changed in a3780cc: hard and soft dependencies now behave like above. Weakdep has no effect at all on modprobe so it's ignored, but since it would fall into not required category it would behave like softdep.

danielzgtg commented on libkmod/libkmod-module.c:
```diff
@@ -1020,6 +1037,9 @@ KMOD_EXPORT int kmod_module_probe_insert_module(
 	}
 
+	if (module_is_masked(mod))
+		return 0;
```
Loading a masked module should not return success.

danielzgtg replied:
Compare systemd:
```console
root@daniel-desktop3:~# systemctl start badservice
Failed to start cron.service: Unit badservice.service is masked.
root@daniel-desktop3:~# echo $?
1
root@daniel-desktop3:~#
```

jspcki replied:
Changed in a3780cc, see full description [above](https://github.com/kmod-project/kmod/pull/312#discussion_r2062591008).

whythefyouflying commented on libkmod/docs/libkmod-docs.xml:
```diff
@@ -60,4 +60,8 @@
     <title>Index of new symbols in 33</title>
```
done

jspcki commented:
> Putting `/bin/true` is the same as putting nothing.
Incorrect.

jspcki commented on libkmod/libkmod-module.c:
```diff
@@ -759,9 +777,21 @@ KMOD_EXPORT int kmod_module_apply_filter(const struct kmod_ctx *ctx,
+			if (mod->required) {
+				ERR(mod->ctx, "module is masked: %s\n",
+				    kmod_module_get_name(mod));
+				err = -EINVAL;
```
I'd like to hear opinions/suggestions on EINVAL in this case and in 1080. I thought about using custom error, but I disliked the part that it "only so happened to" not collide with errno. Otherwise we could return positive error (even though it's not a direct result of flags per `kmod_module_probe_insert_module` description) and handle it in modprobe.c.

jspcki commented on libkmod/libkmod-module.c:
```diff
@@ -759,9 +777,21 @@ KMOD_EXPORT int kmod_module_apply_filter(const struct kmod_ctx *ctx,
 
+		if ((filter_type & KMOD_FILTER_MASK) && module_is_masked(mod)) {
+			if (mod->required) {
+				ERR(mod->ctx, "module is masked: %s\n",
```
We could s/module/dependency/ if it feels more right. It will never be the module itself because of earlier check in 1074 unless used in different conditions.

jspcki commented on tools/modprobe.c:
```diff
@@ -163,6 +163,7 @@ static int show_config(struct kmod_ctx *ctx)
 		const char *name;
 		struct kmod_config_iter *(*get_iter)(const struct kmod_ctx *ctx);
 	} ci[] = {
+		{ "mask", kmod_config_get_masks },
```
This one is not done.

whythefyouflying replied:
> I'm not sure I follow how that makes a difference. People/tools already need to parse the whole output, regardless of how long/short respective sections are.
> 
> Mind you, the suggestion is a bit bike-sheddy so don't take it too seriously.

Alpha sorted everything across the patch **except** this line.

I believe from the human user perspective, given mask is the most restrictive option, it would be good UI to have it displayed first. If we want to alpha sort everything in `show_config()` then we should also print `alias` first and that is not the case.

evelikov reviewed (approved):
Left a few minor comments - with or without them the PR looks good. Thumbs up from me.

Would be interesting to hear Lucas' POV wrt the EINVAL et al.

jslepeck committed:
testsuite: test-modprobe: modprobe with blacklist

Until now blacklist command was only covered by test-blacklist at
libkmod interface level. A regular user will usually interact with it
via modprobe(8). This commit adds tests for modprobe, alias filtering,
and -b option. It is also intended as an example of -b usage.

Co-authored-by: Dawid Osuchowski <dawid.osuchowski@linux.intel.com>
Signed-off-by: Dawid Osuchowski <dawid.osuchowski@linux.intel.com>
Signed-off-by: Jakub Ślepecki <jakub.slepecki@intel.com>
Signed-off-by: Lucas De Marchi <demarchi@kernel.org>

jslepeck committed:
libkmod: unref list before returning with error

In case of error path this might have led to, for example:

    ==86629==ERROR: LeakSanitizer: detected memory leaks

    Direct leak of 24 byte(s) in 1 object(s) allocated from:
        #0 0x7fa904b20e15  (/usr/lib/gcc/x86_64-pc-linux-gnu/15.1.1/../../../../lib/libasan.so+0x120e15) (BuildId: 9dccfc42ea2d9790ff7cf76cc8a3210650e604e0)
        #1 0x55d74820fc8a  ($kmod/build/kmod+0xefc8a) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)
        #2 0x55d7482173a6  ($kmod/build/kmod+0xf73a6) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)
        #3 0x55d74821a617  ($kmod/build/kmod+0xfa617) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)
        #4 0x55d7481e2d80  ($kmod/build/kmod+0xc2d80) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)
        #5 0x55d7481e314f  ($kmod/build/kmod+0xc314f) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)
        #6 0x55d7481e658c  ($kmod/build/kmod+0xc658c) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)
        #7 0x7fa9038376b4  (/usr/lib/libc.so.6+0x276b4) (BuildId: 468e3585c794491a48ea75fceb9e4d6b1464fc35)
        #8 0x7fa903837768  (/usr/lib/libc.so.6+0x27768) (BuildId: 468e3585c794491a48ea75fceb9e4d6b1464fc35)
        #9 0x55d7481c3334  ($kmod/build/kmod+0xa3334) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)

    Indirect leak of 4120 byte(s) in 1 object(s) allocated from:
        #0 0x7fa904b2073d  (/usr/lib/gcc/x86_64-pc-linux-gnu/15.1.1/../../../../lib/libasan.so+0x12073d) (BuildId: 9dccfc42ea2d9790ff7cf76cc8a3210650e604e0)
        #1 0x55d7481eb8eb  ($kmod/build/kmod+0xcb8eb) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)
        #2 0x55d7481f6fee  ($kmod/build/kmod+0xd6fee) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)
        #3 0x55d7481e52c4  ($kmod/build/kmod+0xc52c4) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)
        #4 0x7fa9038376b4  (/usr/lib/libc.so.6+0x276b4) (BuildId: 468e3585c794491a48ea75fceb9e4d6b1464fc35)
        #5 0x7fa903837768  (/usr/lib/libc.so.6+0x27768) (BuildId: 468e3585c794491a48ea75fceb9e4d6b1464fc35)
        #6 0x55d7481c3334  ($kmod/build/kmod+0xa3334) (BuildId: e7c136c95271cf020398e33fd5577d82adc7a47c)

    [...]

    SUMMARY: AddressSanitizer: 5701 byte(s) leaked in 27 allocation(s).

Co-authored-by: Dawid Osuchowski <dawid.osuchowski@linux.intel.com>
Signed-off-by: Dawid Osuchowski <dawid.osuchowski@linux.intel.com>
Signed-off-by: Jakub Ślepecki <jakub.slepecki@intel.com>
Signed-off-by: Lucas De Marchi <demarchi@kernel.org>

osuchowski committed:
libkmod: add mask command

Mask command prevents modules from loading. It intends to be a strict
and intent-based replacement for variations of install and blacklist
commands, for example:

* install module /bin/false
* install module /bin/true
* blacklist module, with modprobe -b

This commit is a pure extension and the already established behaviour
is not expected to change as a result.

Closes: https://github.com/kmod-project/kmod/issues/440
Closes: https://github.com/kmod-project/kmod/issues/40
Co-authored-by: Jakub Ślepecki <jakub.slepecki@intel.com>
Signed-off-by: Jakub Ślepecki <jakub.slepecki@intel.com>
Signed-off-by: Dawid Osuchowski <dawid.osuchowski@linux.intel.com>

whythefyouflying commented on libkmod/libkmod-module.c:
```diff
@@ -1036,15 +1071,23 @@ KMOD_EXPORT int kmod_module_probe_insert_module(
 		if (err < 0)
 			return err;
 
-		kmod_module_unref_list(list);
```
Done

jspcki replied:
I observed a crash primarily at 1072 when it had similar sequence of statements -> filter -> return -> unref. You'd need to hit any of the failing conditions of the filter: `(ctx == NULL || output == NULL)`, `mod->required`, or `kmod_list_append(*output, mod) == NULL`. Other than `required` they are unlikely. Right now easiest way would be to inject error there manually and it's the way I generated sanitizer output for the commit.

whythefyouflying commented:
We have rebased the code to latest master and addressed review concerns

whythefyouflying commented:
@lucasdemarchi Can you please take a look?

lucasdemarchi committed:
libkmod: Fix missing module normalization from kernel cmdline

Normalize it before adding the blacklist. Without normalizing it we
have, with just the test addition:

	$ ./build/testsuite/test-modprobe modprobe_param_kcmdline9
	TESTSUITE: running modprobe_param_kcmdline9, in forked context
	TESTSUITE: ERR: Outputs do not match on STDOUT:
	TESTSUITE: ERR: correct:
	...
	blacklist foo_bar
	...
	TESTSUITE: ERR: wrong:
	blacklist foo-bar
	TESTSUITE: ERR: FAILED: exit ok but outputs do not match: modprobe_param_kcmdline9

Signed-off-by: Lucas De Marchi <demarchi@kernel.org>

lucasdemarchi commented:
I rebased this PR and solved a few small issues. I also added a fix to kcmdline parsing for blacklist that I noticed while reviewing/rebase.

lucasdemarchi commented:
The first 3 commits should be good to go. The last one I will take another look on the next days. Sorry for the delay and thanks for adding this.

lucasdemarchi committed:
Fix some trivial coding style issues

Make clang-format happy.

Signed-off-by: Lucas De Marchi <demarchi@kernel.org>
